src/target/adiv5_swdp.c

```diff
@@ -93,93 +97,98 @@ int adiv5_swdp_scan(uint32_t targetid)
 			initial_dp->seq_out(0xE79E, 16); /* 0b0111100111100111 */
+
 			dp_line_reset(initial_dp);
+
 			initial_dp->fault = 0;
 
 			TRY_CATCH (e, EXCEPTION_ALL) {
-				initial_dp->debug_port_id = initial_dp->dp_read(initial_dp, ADIV5_DP_DPIDR);
+				dp_dpidr = initial_dp->dp_read(initial_dp, ADIV5_DP_DPIDR);
 			}
 			if (e.type || initial_dp->fault) {
 				DEBUG_WARN("No usable DP found\n");
 				return -1;
 			}
 		}
 
-		initial_dp->version =
-			(initial_dp->debug_port_id & ADIV5_DP_DPIDR_VERSION_MASK) >> ADIV5_DP_DPIDR_VERSION_OFFSET;
-		if (initial_dp->version >= 2) {
+		const uint8_t dp_version = (dp_dpidr & ADIV5_DP_DPIDR_VERSION_MASK) >> ADIV5_DP_DPIDR_VERSION_OFFSET;
+		if (dp_version >= 2) {
 			scan_multidrop = true;
+
 			/* Read TargetID. Can be done with device in WFI, sleep or reset!*/
 			adiv5_dp_write(initial_dp, ADIV5_DP_SELECT, 2); /* TARGETID is on bank 2 */
-			initial_dp->target_id = adiv5_dp_read(initial_dp, ADIV5_DP_TARGETID);
+			dp_targetid = adiv5_dp_read(initial_dp, ADIV5_DP_TARGETID);
 			adiv5_dp_write(initial_dp, ADIV5_DP_SELECT, 0);
 
 			const uint16_t tdesigner =
-				(initial_dp->target_id & ADIV5_DP_TARGETID_TDESIGNER_MASK) >> ADIV5_DP_TARGETID_TDESIGNER_OFFSET;
-			const uint16_t tpartno =
-				(initial_dp->target_id & ADIV5_DP_TARGETID_TPARTNO_MASK) >> ADIV5_DP_TARGETID_TPARTNO_OFFSET;
+				(dp_targetid & ADIV5_DP_TARGETID_TDESIGNER_MASK) >> ADIV5_DP_TARGETID_TDESIGNER_OFFSET;
+			const uint16_t tpartno = (dp_targetid & ADIV5_DP_TARGETID_TPARTNO_MASK) >> ADIV5_DP_TARGETID_TPARTNO_OFFSET;
 			/* convert it to our internal representation, See JEP-106 code list */
 			const uint16_t designer_code = (tdesigner & ADIV5_DP_DESIGNER_JEP106_CONT_MASK) << 1U |
 			                               (tdesigner & ADIV5_DP_DESIGNER_JEP106_CODE_MASK);
-			if (designer_code == JEP106_MANUFACTURER_RASPBERRY && tpartno == 0x2) {
+
+			if (designer_code == JEP106_MANUFACTURER_RASPBERRY && tpartno == 0x1002) {
 				/* RP2040 */
 				/* Release evt. handing RESCUE DP reset*/
-				adiv5_dp_write(initial_dp, ADIV5_DP_CTRLSTAT, 0);
-			}
 
-			if (!initial_dp->dp_low_write) {
-				DEBUG_WARN("CMSIS_DAP < V1.2 can not handle multi-drop!\n");
-				/* E.g. CMSIS_DAP < V1.2 can not handle multi-drop!*/
-				scan_multidrop = false;
+				adiv5_dp_write(initial_dp, ADIV5_DP_CTRLSTAT, 0);
 			}
-		} else {
-			scan_multidrop = false;
 		}
-	} else {
-		initial_dp->target_id = targetid;
 	}
-	const volatile size_t nr_dps = (scan_multidrop) ? 16U : 1U;
-	volatile uint32_t dp_targetid;
-	volatile uint32_t dp_debug_port_id;
-	for (volatile size_t i = 0; i < nr_dps; i++) {
+
+	if (!initial_dp->dp_low_write) {
+		DEBUG_WARN("CMSIS_DAP < V1.2 can not handle multi-drop!\n");
+		/* E.g. CMSIS_DAP < V1.2 can not handle multi-drop!*/
+		scan_multidrop = false;
+	}
+
+	DEBUG_WARN("scan_multidrop: %s\n", scan_multidrop ? "true" : "false");
+
+	const volatile size_t max_dp = (scan_multidrop) ? 16U : 1U;
+	for (volatile size_t i = 0; i < max_dp; i++) {
```

Comment:
imo it should not be required, but I was weary to remove previous volatiles

Reply:
To be honest.. if volatile is the only think making that loop work, we've got much bigger problems - there's nothing about it that should need volatile. It looks like it was added "in case" the compiler tried optimising the loop out, but the compiler won't as it makes calls and does work

Reply:
like `ctrlstat`, removing volatile triggers `error: variable ‘i’ might be clobbered by ‘longjmp’ or ‘vfork’ [-Werror=clobbered]`

Reply:
Ok, that's a good explanation.. it indicates an underlying bigger problem, but we'll take it for now. Please add a note about this to the definition.